src/portgraph.rs

```diff
@@ -671,6 +678,90 @@ impl PortGraph {
         self.port_link.reserve(ports);
     }
 
+    /// Changes the number of ports of the `node` to the given `incoming` and `outgoing` counts.
+    ///
+    /// Invalidates the indices of the node's ports. If the number of incoming or outgoing ports
+    /// is reduced, the ports are removed from the end of the port list.
+    ///
+    /// Every time a port is moved, the `rekey` function will be called with its old and new index.
+    /// If the port is removed, the new index will be `None`.
+    ///
+    /// This operation is O(n) where n is the number of ports of the node.
+    pub fn set_num_ports<F>(
+        &mut self,
+        node: NodeIndex,
+        incoming: usize,
+        outgoing: usize,
+        mut rekey: F,
+    ) where
+        F: FnMut(PortIndex, Option<PortIndex>),
+    {
+        // TODO: Add port capacity and use a grow factor to avoid unnecessary reallocations.
+
+        assert!(incoming < u16::MAX as usize - 1);
+        assert!(outgoing < u16::MAX as usize - 1);
```

Comment:
`NodeMeta` uses a `NonZeroU16` to count the number of incoming ports, so that `NodeEntry` can be efficiently encoded. I'll add a `NodeMeta::MAX_{INCOMING,OUTGOING}` to make this clearer.